enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,476 @@
+---
+title: forwarding-json-structured-logs
+
+authors:
+  - "@alanconway"
+
+reviewers:
+  - "@ewolinetz"
+  - "@lukas-vlcek"
+  - "@jcantril"
+  - "@sichvoge"
+
+approvers:
+  - "@jcantril"
+
+creation-date: 2020-10-27
+
+status: provisional
+
+see-also:
+  - ./cluster-logging-es-pipeline-processing.md
+  - ./forwarder-label-selector.md
+---
+
+[es-pipeline]: ./cluster-logging-es-pipeline-processing.md "Elasticsearch Pipeline Processing"
+[clf-labels]: ./forwarder-label-selector.md
+[es-issue]: https://github.com/openshift/origin-aggregated-logging/issues/1492
+[defunct_merge]: https://docs.openshift.com/container-platform/4.1/logging/config/efk-logging-fluentd.html#efk-logging-fluentd-json_efk-logging-fluentd "Defunct `MERGE_JSON_LOG` feature"
+[data_model]: https://github.com/openshift/origin-aggregated-logging/blob/master/docs/com.redhat.viaq-openshift-project.asciidoc
+[k8s_labels]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels
+[recommended_labels]: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/ "Recommended k8s labels"
+[jsonpath]: https://goessner.net/articles/JsonPath/
+
+# Forwarding JSON Structured Logs
+
+## Release Sign-off Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+When applications write structured JSON logs, consumers want to access fields of JSON log entries for indexing and other purposes.
+The current logging [data model][data_model] stores the log entry as a JSON *string*, not a JSON *object*.
+Consumers can't access the log entry fields without a second JSON parse of this string.
+
+The current implementation also 'flattens' labels to work around Elasticsearch limitations.
+
+To illustrate, given a log entry `{"name":"fred","home":"bedrock"}` from a container with the label `app.kubernetes.io/name="flintstones"`. The current output record is:
+
+```json
+{
+  "message":"{\"name\":\"fred\",\"home\":\"bedrock\"}",
+  "kubernetes":{"labels":{"app_kubernetes_io/name": "flintstones", ...}},
+  ... other metadata
+}
+```
+
+This proposal will also allow alternate forms of output record such as:
+
+Complete record, compliant with the new logging [data model][model]
+```json
+{
+  "structured":{"name":"fred","home":"bedrock"}",
+  "kubernetes":{"labels":{"app.kubernetes.io/name": "flintstones", ...}},
+  ... other metadata
+}
+```
+
+Merged record with user fields at top level, restores lost [`MERGE_JSON_LOG`][defunct_merge] feature of OCP 4.1.
+``` json
+{
+  "name":"fred",
+  "home":"bedrock",
+  "kubernetes":{"labels":{"app.kubernetes.io/name": "flintstones", ...}},
+  ... other metadata
+}
+```
+
+Customized record with cherry-picked meta-data
```

Comment:
these are three different possibilities?
if so, can we make this into a numbered list or something to make this more obvious?

Reply:
They are 2 possibilities, the first one has the user JSON in the `structured` field, the second has "merged" the user fields directly into our envelope. There's debate about whether we should even allow that - the motivation is to match our defunct MERGE_JSON_LOG feature, but it's dangerous (name conflicts) so if we don't have to support that we can get rid of the second option. 
@sichvoge any thoughs on that?

Reply:
I'll update the wording anyway to be clearer.